rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,421 @@
+addressing the issues enumerated above. The core changes to be made are as
+follows:
+
+ - Make `Parent` and `Children` (mostly) publicly immutable.
+ - Make mutations to the hierarchy rely solely on commands.
+ - Update hierarchy commands to automatically update both `Parent` and
+   `Children` to remain consistent with each other.
+ - Remove the `parent_update_system`.
+ - Remove `PreviousParent`.
+ - Register several hierarchy events that fires whenever time a new child is
+   added or removed from a parent.
+ - Create a `HiearchyQuery` custom SystemParam for more easily iterating over
+   components in the hierarchy.
+
+This change will repurpose the following existing `EntityCommand` extension
+functions for this purpose:
+
+ - `add_child`
+ - `push_children`
+ - `insert_children`
+ - `remove_child` (this will be added as a wrapper around `remove_children`)
+ - `remove_children`
+
+An additional `set_parent(Option<Entity>)` command should also be added as a
+child-side way of altering the hierarchy, defaulting to adding the entity as the
+last child to the provided parent if provided, and detaching the child from it's
+parent if `None`.
+
+### HierarchyQuery
+An extension to this design is to create dedicated `HierarchyQuery<Q, F>`, which
+is a custom `SystemParam` that wraps multiple `Query` objects to simplify
+querying for items in the hierarchy:
+
+ - `roots(_mut)` - returns an iterator over entities that match the provided
+   query parameters that do not have a parent.
+ - `parent(_mut)` - returns single result from the parent of an entity.
+ - `iter_children(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does not return Children of Children.
+ - `iter_descendants(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This does a depth-first traversal over the hierarchy
+   starting at some entity.
+ - `iter_ancestors(_mut)` - returns an iterator over `(parent, child)` results
+   from a query parameter. This bubbles up queries from the bottom to the top
+   starting at some entity.
+
+As suggested by the generic parameters, the query results and filters are
+identical to the ones used by a normal Query or QueryState.
+
+The main motivation for this is to make creating systems like
+`transform_propagate_system` easier. These queries can be generically used for
+any system that needs some form of hierachy traversal. These queries can be quite
+expensive to run, so the documentation for them should reflect this and
+discourage misuing it.
+
+### HierarchyEvents
+Three new events should be added here, each corresponding to a specific change
+in the hierarchy.
+
+```rust
+pub enum ChildAdded {
+  pub parent: Entity,
+  pub child: Entity,
+}
+
+pub enum ChildRemoved {
+  pub parent: Entity,
+  pub child: Entity,
+}
+
+pub enum ChildMoved {
+  pub parent: Entity,
+  pub previous_parent: Entity,
+  pub new_parent: Entity,
+}
```

Comment:
```suggestion
pub enum HierarchyEvent {
  ChildAdded {
    parent: Entity,
    child: Entity,
  },
  ChildRemoved {
    parent: Entity,
    child: Entity,
  },
  ChildMoved {
    parent: Entity,
    previous_parent: Entity,
    new_parent: Entity,
  },
}
```

You probably meant this, right?

Reply:
The intention is to keep them separate so listeners can independently listen to each kind of event separately instead of listening on instances of all 3. Plus moving a child suggests the other two.

Reply:
The downside to separate queues here is that order matters for these events. ChildRemoved -> ChildAdded is different from ChildAdded -> ChildRemoved. If the goal is to maintain some consistent "view" of the hierarchy, I think we need a centralized queue.

I'd also like some clarity on when ChildMoved vs ChildRemoved+ChildAdded events will be sent.

Reply:
Great point, updated to match this. Also this should really simplify the event code (looking at the prototype PR) haha.